crates/bevy_render/src/batching/gpu_preprocessing.rs

```diff
@@ -76,10 +129,114 @@ pub struct PreprocessWorkItem {
+///     first_instance: u32,
+/// }
+/// ```
+///
+/// We actually generally treat these two variants identically in code. To do
+/// that, we make the following two observations:
+///
+/// 1. `instance_count` is in the same place in both structures. So we can
+/// access it regardless of the structure we're looking at.
+///
+/// 2. The second structure is one word larger than the first. Thus we need to
+/// pad out the first structure by one word in order to place both structures in
+/// an array. If we pad out `ArrayIndirectParameters` by copying the
+/// `first_instance` field into the padding, then the resulting union structure
+/// will always have a read-only copy of `first_instance` in the final word. We
+/// take advantage of this in the shader to eliminate branching.
+#[derive(Clone, Copy, Pod, Zeroable, ShaderType)]
+#[repr(C)]
+pub struct IndirectParameters {
+    /// For `ArrayIndirectParameters`, `vertex_count`; for
+    /// `ElementIndirectParameters`, `index_count`.
+    pub data0: u32,
+
+    /// The number of instances we're going to draw.
+    ///
+    /// This field is in the same place in both structures.
+    pub instance_count: u32,
+
+    /// The index of the first vertex we're to draw.
+    pub first_vertex: u32,
+
+    /// For `ArrayIndirectParameters`, `first_instance`; for
+    /// `ElementIndirectParameters`, `base_vertex`.`
+    pub data1: u32,
+
+    /// For `ArrayIndirectParameters`, this is padding; for
+    /// `ElementIndirectParameters`, this is `first_instance`.
+    ///
+    /// Conventionally, we copy `first_instance` into this field when padding
+    /// out `ArrayIndirectParameters`. That way, shader code can read this value
+    /// at the same place, regardless of the specific structure this represents.
+    pub first_instance: u32,
+}
+
+/// The buffer containing the list of [`IndirectParameters`], for draw commands.
+#[derive(Resource, Deref, DerefMut)]
+pub struct IndirectParametersBuffer(pub BufferVec<IndirectParameters>);
+
+impl IndirectParametersBuffer {
+    /// Creates the indirect parameters buffer.
+    pub fn new() -> IndirectParametersBuffer {
+        IndirectParametersBuffer(BufferVec::new(
+            BufferUsages::STORAGE | BufferUsages::INDIRECT,
+        ))
+    }
+}
+
+impl Default for IndirectParametersBuffer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FromWorld for GpuPreprocessingSupport {
+    fn from_world(world: &mut World) -> Self {
+        let adapter = world.resource::<RenderAdapter>();
+        let device = world.resource::<RenderDevice>();
+
+        if device.limits().max_compute_workgroup_size_x == 0 {
+            GpuPreprocessingSupport::None
+        } else if !device
+            .features()
+            .contains(Features::INDIRECT_FIRST_INSTANCE) ||
+            !adapter.get_downlevel_capabilities().flags.contains(
+        DownlevelFlags::VERTEX_AND_INSTANCE_INDEX_RESPECTS_RESPECTIVE_FIRST_VALUE_IN_INDIRECT_DRAW)
```

Comment:
Just noting that I was going to suggest adding a comment that this checks for a DX12 issue, but that feels like the kind of comment that becomes outdated. I hope one can look at the documentation of the downlevel flag and get more and up to date information about it.

Reply:
This solution also means it will automatically work once wgpu has fixed the issue without us needing to change any code.